### PR TITLE
Add Index Encoding code from backend

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds the Index Encoding code that the backend uses to encode field values to index entries. Datastore-only features are removed and otherwise unused methods were removed.

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/index/DirectionalIndexByteEncoder.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.index;
+
+import com.google.protobuf.ByteString;
+
+/** An index value encoder. */
+public abstract class DirectionalIndexByteEncoder {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/index/DirectionalIndexByteEncoder.java:20]: https://cs/piper///depot/google3/java/com/google/cloud/datastore/encoding/StorageCodecs.java;l=17?q=DirectionalIndexByteEncoder

File: firebase-firestore/src/main/java/com/google/firebase/firestore/index/FirestoreIndexValueWriter.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import java.util.Map;
+
+/**
+ * Firestore index value writer.
+ *
+ * <p>See the <a
+ * href="https://g3doc.corp.google.com/cloud/datastore/g3doc/architecture/spanner/storage_format.md">storage
+ * format design</a> for details.
+ */
+public class FirestoreIndexValueWriter {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/index/FirestoreIndexValueWriter.java:32]: cs/piper///depot/google3/java/com/google/cloud/datastore/encoding/FirestoreIndexValueWriter.java

File: firebase-firestore/src/main/java/com/google/firebase/firestore/index/IntMath.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * Jr.'s <i>Hacker's Delight</i>, (Addison Wesley, 2002).
+ *
+ * <p>Similar functionality for {@code long} and for {@link BigInteger} can be found in {@link
+ * LongMath} and {@link BigIntegerMath} respectively. For other common operations on {@code int}
+ * values, see {@link com.google.common.primitives.Ints}.
+ *
+ * @author Louis Wasserman
+ * @since 11.0
+ */
+public final class IntMath {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/index/IntMath.java:37]: cs/third_party/java_src/google_common/current/java/com/google/common/math/IntMath.java

File: firebase-firestore/src/main/java/com/google/firebase/firestore/index/OrderedCodeReader.java [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+import static java.lang.Character.MIN_SUPPLEMENTARY_CODE_POINT;
+import static java.lang.Character.MIN_SURROGATE;
+
+/**
+ * OrderedCodeReader is a minimal-allocation implementation of the reading behavior defined by
+ * {@link com.google.bigtable.util.OrderedCode}.
+ *
+ * <p>This class will throw {@link IllegalArgumentException} on invalid input.
+ */
+public class OrderedCodeReader {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/index/OrderedCodeReader.java:34]: cs/piper///depot/google3/java/com/google/cloud/datastore/encoding/orderedcode/OrderedCodeReader.java

File: firebase-firestore/src/main/java/com/google/firebase/firestore/index/OrderedCodeWriter.java [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+import com.google.protobuf.ByteString;
+import java.math.RoundingMode;
+import java.util.Arrays;
+
+/**
+ * OrderedCodeWriter is a minimal-allocation implmentation of the writing behavior defined by {@link
+ * com.google.bigtable.util.OrderedCode}.
+ */
+public class OrderedCodeWriter {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/index/OrderedCodeWriter.java:28]: cs/piper///depot/google3/java/com/google/cloud/datastore/encoding/orderedcode/OrderedCodeReader.java

**Review by thebrianchen — APPROVED** [review]

**Pull request merged**